README.md
# jekyll-postcss-browserify

[![js-standard-style](https://img.shields.io/badge/code%20style-standard-brightgreen.svg)](http://standardjs.com/) [![Build Status](https://travis-ci.org/cbracco/jekyll-postcss-browserify.svg)](https://travis-ci.org/cbracco/jekyll-postcss-browserify)

> A boilerplate for building static websites with [Gulp][gulp], [Jekyll][jekyll], [PostCSS][postcss], [Browserify][browserify], [imagemin][imagemin], [svg-sprite][svg-sprite], and more.

This project is an opinionated front-end boilerplate for developing and deploying static websites using [Jekyll][jekyll]. It includes several front-end tools that help facilitate the development process, and [Gulp][gulp] to automate common development tasks, compiling source code, watching files for changes, and deployment.

- [Features](#features)
- [Requirements](#requirements)
- [Installation](#installation)
- [Usage](#usage)
  - [Developing](#developing)
  - [Publishing](#publishing)
  - [Deploying](#deploying)
- [Credits](#credits)
- [Contributing](#contributing)
- [Changelog](#changelog)
- [License](#license)

## Features

- [Gulp][gulp] as a build system to automate tasks, compile source code, watch files for changes, and deployment.
- [Jekyll][jekyll] as a blog-focused static site generator that uses the [Liquid][liquid] templating language.
- [PostCSS][postcss] for CSS [file imports][postcss-import], [url rebasing][postcss-url], [future-friendly syntax, vendor prefixing][postcss-cssnext], and more.
- [Browserify][browserify] for requiring JavaScript modules and bundling dependencies in the browser.
- [imagemin][imagemin] for compressing image files.
- [svg-sprite][svg-sprite] to automatically generate and optimize an SVG sprite from raw SVG icon files.
- [Browsersync][browsersync] with [xip.io][xip.io] for serving files, live-reloading browsers, and multi-device testing.

## Requirements

This project is compatible with **Linux**, **Unix**, and **Mac OSX** operating systems. It requires the following software to be already installed on your system:

- [Ruby][ruby] (>= 2.3.0)
- [RubyGems][rubygems] (>= 2.6.11)
- [Bundler][rubybundler] (>= 1.14.6)
- [Node.js][node] (>= 8.11.3)
- [npm][npm] (>= 5.0.0)

**Running Windows?** Jekyll does not officially support the Windows operating system, but they do provide [special instructions][jekyll-windows] if you want to take a crack at it anyway.

## Installation

Clone this repository to your local machine, and install its dependencies with the following command(s):

```bash
git clone git@github.com:cbracco/jekyll-postcss-browserify.git
cd jekyll-postcss-browserify/
npm install
```

The `npm install` command will first run `bundle install` to install the required Ruby gems, and then `npm install` for the required npm modules.

## Usage

### Developing

To start working, first run the following command(s):

```bash
npm start
```

The `npm start` command will run the default Gulp task which builds the Jekyll templates and compiles the website assets, starts a local server (http://127.0.0.1.xip.io:9999) with Browsersync, opens the project in the default browser, and begins watching for file changes.

This is considered **“Development Mode”**, which means that the files being served in this local environment are not yet minified or optimized for deployment.

### Publishing

To publish a production-ready version of the website, run the following command(s):

```bash
npm run production
```

The `npm run production` command will run the production Gulp task which deletes any existing production builds, builds a new production-ready version of the website, starts a local server (http://127.0.0.1.xip.io:9998) with Browsersync, and opens the project in the default browser.

This can be considered **“Production Mode”**, which means that the files being served in this local environment are minified and optimized for deployment. It is intended to mimic the production environment on your local environment, so you can preview exactly what you will see before you deploy the website to a production server.

### Deploying

Before deploying the website, first create a file called `rsync-credentials.json` and fill it with the following information:

```json
{
  "destination": "/path/to/your/website",
  "hostname": "example.com",
  "username": "person",
  "port": 666
}
```

Then, run the following command(s) to deploy a production-ready version of the website to a remote server:

```bash
npm run deploy
```

The `npm run deploy` command will run the deploy Gulp task which deletes any existing production builds, builds a new production-ready version of the website, and uses [rsync][rsync] to upload the website to the specified remote server.

## Credits

> *“If I have seen further it is by standing on ye sholders of Giants.”*
> &mdash; [Issac Newton][issac-newton-quote]

jekyll-postcss-browserify is a project by [@cbracco][cbracco] and its [contributors][contributors]. It is influenced by the following organizations, projects, articles, and individuals:

- [gulp-tutorial][gulp-tutorial] by [@kogakure][kogakure]
- [tape-standard][tape-standard] by [@rstacruz][rstacruz]

Thank you all for your hard work.

## [Contributing](CONTRIBUTING.md)

## [Changelog](CHANGELOG.md)

## [License](LICENSE)

[browserify]: http://browserify.org
[browsersync]: https://browsersync.io
[cbracco]: https://github.com/cbracco
[contributors]: https://github.com/cbracco/jekyll-postcss-browserify/graphs/contributors
[gulp]: http://gulpjs.com
[gulp-tutorial]: https://github.com/kogakure/gulp-tutorial
[imagemin]: https://github.com/imagemin/imagemin
[issac-newton-quote]: https://en.wikipedia.org/wiki/Standing_on_the_shoulders_of_giants
[jekyll]: https://jekyllrb.com
[jekyll-windows]: http://jekyllrb.com/docs/windows/#installation
[kogakure]: https://github.com/kogakure
[liquid]: http://liquidmarkup.org
[node]: https://nodejs.org/en/
[npm]: https://www.npmjs.com
[postcss]: http://postcss.org
[postcss-cssnext]: https://github.com/MoOx/postcss-cssnext
[postcss-import]: https://github.com/postcss/postcss-import
[postcss-url]: https://github.com/postcss/postcss-url
[rstacruz]: https://github.com/rstacruz
[rsync]: http://linux.die.net/man/1/rsync
[ruby]: https://www.ruby-lang.org/en/
[rubybundler]: http://bundler.io
[rubygems]: https://rubygems.org
[svg-sprite]: https://github.com/jkphl/svg-sprite
[tape-standard]: https://github.com/rstacruz/tape-standard
[xip.io]: http://xip.io


CHANGELOG.md
# 0.1.1 (December 28, 2019)

- Fix tests

# 0.1.0 (January 2, 2018)

- Update to Gulp 4.
- Update dependencies.
- Re-write Gulp tasks using new syntax (series, parallel, etc.).
- Add `gulp-size` to styles tasks.
- Update README.
- Update npm scripts.
- Update Travis CI build instructions.
- Update tests.
- Add package-lock.json file.
- Rename `tools/` directory to `gulpfile.js/`.

# 0.0.5 (February 21, 2016)

- Add Gulp watch tasks for icons.

# 0.0.4 (February 21, 2016)

- Add empty `/icons` folder and update Gulp config.

# 0.0.3 (February 13, 2016)

- Update travis-ci scripts and configuration.
- Fix typos in README.

# 0.0.2 (February 13, 2016)

- Add necessary scripts for travis-ci and update gems.

# 0.0.1 (February 13, 2016)

- Initial release.


gulpfile.js/config.js
// Define Paths
const src = './source/'
const srcAssets = './source/_assets/'
const build = './build/'
const development = './build/development/'
const developmentAssets = './build/assets/'
const production = './build/production/'
const productionAssets = './build/production/assets/'

// Require rsyncCredentials only if the file exists to avoid build
// failure on travis-ci
const fs = require('fs')
let rsyncCredentials

if (fs.existsSync('../../rsync-credentials.json')) {
  rsyncCredentials = require('../../rsync-credentials.json')
} else {
  rsyncCredentials = {
    destination: '/path/to/website',
    hostname: 'example.com',
    username: 'user',
    port: 666
  }
}

// Define Configuration
module.exports = {
  watch: {
    jekyll: [
      '_config.yml',
      '_config.production.yml',
      src + '_data/**/*',
      src + '_includes/**/*',
      src + '_layouts/**/*',
      src + '_plugins/**/*',
      src + '_posts/**/*',
      src + '**/*.{html,markdown,md,yml,json,txt,xml}',
      src + '*'
    ],
    styles: srcAssets + 'styles/**/*',
    scripts: srcAssets + 'scripts/**/*',
    images: srcAssets + 'images/**/*',
    icons: srcAssets + 'icons/**/*',
    favicons: srcAssets + 'favicons/**/*',
    fonts: srcAssets + 'fonts/**/*'
  },
  serve: {
    development: {
      server: {
        baseDir: [development, build, src]
      },
      files: [
        developmentAssets + '/styles/*.css',
        developmentAssets + '/scripts/*.js',
        developmentAssets + '/images/**',
        developmentAssets + '/icons/**',
        developmentAssets + '/favicons/*'
      ],
      port: 9999,
      xip: true,
      notify: {
        styles: ['display: hidden; padding: 12px; font-family: sans-serif; position: fixed; font-size: 14px; z-index: 10000; left: 0; bottom: 0; width: 200px; margin: 0; border-top-left-radius: 0; border-top-right-radius: 3px; color: #fff; text-align: center; background-color: rgba(0, 0, 0, 0.75);']
      }
    },
    production: {
      server: {
        baseDir: [production]
      },
      port: 9998,
      xip: true
    }
  },
  build: {},
  clean: {
    src: developmentAssets
  },
  jekyll: {
    development: {
      src: src,
      dest: development,
      config: '_config.yml'
    },
    production: {
      src: src,
      dest: production,
      config: '_config.yml,_config.production.yml'
    }
  },
  styles: {
    src: srcAssets + 'styles/styles.css',
    dest: developmentAssets + 'styles/',
    processors: {
      cssnext: {
        browsers: [
          'last 2 versions'
        ]
      },
      mqpacker: {
        sort: true
      }
    }
  },
  scripts: {
    debug: true,
    extensions: ['.coffee', '.hbs'],
    bundleConfigs: [
      {
        entries: srcAssets + 'scripts/scripts.js',
        dest: developmentAssets + 'scripts/',
        outputName: 'scripts.js'
      },
      {
        entries: srcAssets + 'scripts/head.js',
        dest: developmentAssets + 'scripts/',
        outputName: 'head.js'
      }
    ]
  },
  images: {
    src: srcAssets + 'images/**/*',
    dest: developmentAssets + 'images/'
  },
  icons: {
    src: srcAssets + 'icons/**/*.svg',
    dest: developmentAssets + 'icons',
    processors: {
      sprite: {
        mode: {
          symbol: {
            dest: './',
            sprite: 'sprite.svg'
          }
        }
      }
    }
  },
  favicons: {
    src: srcAssets + 'favicons/**/*',
    dest: developmentAssets + 'favicons/'
  },
  fonts: {
    src: srcAssets + 'fonts/**/*',
    dest: developmentAssets + 'fonts/'
  },
  optimize: {
    styles: {
      src: developmentAssets + '/styles/*.css',
      dest: productionAssets + '/styles/',
      processors: {
        rename: {
          suffix: '.min'
        }
      }
    },
    scripts: {
      src: developmentAssets + '/scripts/*.js',
      dest: productionAssets + '/scripts/',
      processors: {
        rename: {
          suffix: '.min'
        },
        uglify: {}
      }
    },
    images: {
      src: developmentAssets + '/images/**/*.{jpg,jpeg,png,gif}',
      dest: productionAssets + '/images/',
      processors: {
        imagemin: {
          optimizationLevel: 3,
          progessive: true,
          interlaced: true,
          svgoPlugins: [{
            removeViewBox: false
          }]
        }
      }
    },
    icons: {
      src: developmentAssets + '/icons/**/*',
      dest: productionAssets + '/icons/'
    },
    favicons: {
      src: developmentAssets + '/favicons/**/*',
      dest: productionAssets + '/favicons/'
    },
    fonts: {
      src: developmentAssets + '/fonts/**/*',
      dest: productionAssets + '/fonts/'
    },
    html: {
      src: production + '/**/*.html',
      dest: production,
      processors: {
        htmlmin: {
          collapseWhitespace: true
        },
        htmlreplace: {
          styles: {
            src: '/assets/styles/',
            tpl: '<link rel="stylesheet" href="%sstyles.min.css">'
          },
          head: {
            src: '/assets/scripts/',
            tpl: '<script src="%shead.min.js"></script>'
          },
          scripts: {
            src: '/assets/scripts/',
            tpl: '<script src="%sscripts.min.js"></script>'
          }
        }
      }
    }
  },
  revision: {
    src: {
      assets: [
        productionAssets + '/styles/*.css',
        productionAssets + '/scripts/*.js',
        productionAssets + '/images/**/*'
      ],
      base: production
    },
    dest: {
      assets: production,
      manifest: {
        name: 'manifest.json',
        path: productionAssets
      }
    },
    collect: {
      src: [
        productionAssets + '/manifest.json',
        production + '/**/*.{html,xml,txt,json,css,js}',
        '!' + production + '/feed.xml'
      ],
      dest: production
    }
  },
  gzip: {
    src: production + '/**/*.{html,xml,json,css,js}',
    dest: production,
    options: {}
  },
  rsync: {
    src: production + '/**',
    options: {
      destination: rsyncCredentials.destination,
      root: production,
      hostname: rsyncCredentials.hostname,
      username: rsyncCredentials.username,
      port: rsyncCredentials.port,
      incremental: true,
      progress: true,
      relative: true,
      emptyDirectories: true,
      recursive: true,
      clean: true,
      exclude: ['.DS_Store'],
      include: []
    }
  }
}
